spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/MessageConverterConfigurer.java

```diff
@@ -168,12 +159,78 @@ public Object fromMessage(Message<?> message, Class<?> targetClass, Object conve
 		 * @return the converted message
 		 */
 		protected Object build(Object payload, MessageHeaders headers) {
-			MimeType messageContentType = MessageConverterUtils.X_JAVA_OBJECT.equals(contentType) ?
-					MessageConverterUtils.javaObjectMimeType(payload.getClass()) : contentType;
-			return messageBuilderFactory.withPayload(payload).copyHeaders(headers)
+			MimeType messageContentType = MessageConverterUtils.X_JAVA_OBJECT.equals(this.contentType) ?
+					MessageConverterUtils.javaObjectMimeType(payload.getClass()) : this.contentType;
+			return MessageConverterConfigurer.this.messageBuilderFactory.withPayload(payload).copyHeaders(headers)
 					.copyHeaders(Collections.singletonMap(MessageHeaders.CONTENT_TYPE,
 							messageContentType.toString()))
 					.build();
 		}
 	}
+
+	private final class ContentTypeConvertingInterceptor extends ChannelInterceptorAdapter {
+
+		private final String contentType;
+
+		private final MimeType mimeType;
+
+		private ContentTypeConvertingInterceptor(String contentType) {
+			this.contentType = contentType;
+			this.mimeType = MessageConverterUtils.getMimeType(contentType);
+		}
+
+		@Override
+		public Message<?> preSend(Message<?> message, MessageChannel channel) {
+			Class<?>[] classes =
+					MessageConverterConfigurer.this.compositeMessageConverterFactory.supportedDataTypes(
+							this.mimeType);
+			MessageWrappingMessageConverter messageConverter =
```

Comment:
could defer creation of this until the last `else` block below since that's the only place it's used? (no need to create here if the method might `return` without using it)